File: apps/web/server/services/renderHash.ts
import { createHash } from "crypto";
import type {
  VideoEditorProject,
  RenderProfile,
  Clip,
  Track,
} from "../../client/src/types/videoEditor";

export type { RenderProfile } from "../../client/src/types/videoEditor";

interface CanonicalClip {
  assetId: string;
  startTime: number;
  duration: number;
  trimIn: number;
  trimOut: number;
  volume: number;
  speed: number;
  effects: unknown[];
  inTransition?: unknown;
  transform?: unknown;
  textConfig?: unknown;
  transitions?: unknown;
}

function canonicalizeClips(clips: Clip[]): CanonicalClip[] {
  return [...clips]
    .sort((a, b) => a.startTime - b.startTime)
    .map((clip) => {
      const canonical: CanonicalClip = {
        assetId: clip.assetId,
        startTime: clip.startTime,
        duration: clip.duration,
        trimIn: clip.trimIn,
        trimOut: clip.trimOut,
        volume: clip.volume,
        speed: clip.speed,
        effects: clip.effects || [],
      };
      if (clip.inTransition) canonical.inTransition = clip.inTransition;
      if (clip.transform) canonical.transform = clip.transform;
      if (clip.textConfig) canonical.textConfig = clip.textConfig;
      if (clip.transitions) canonical.transitions = clip.transitions;
      return canonical;
    });
}

function canonicalizeTracks(tracks: Track[]) {
  return tracks.map((track) => ({
    type: track.type,
    name: track.name,
    clips: canonicalizeClips(track.clips),
    muted: track.muted,
  }));
}

/**
 * Deterministic JSON serialization with sorted keys and compact separators.
 * Matches Python's json.dumps(obj, sort_keys=True, separators=(',', ':')).
 */
function stableStringify(obj: unknown): string {
  if (obj === null || obj === undefined) return "null";
  if (typeof obj === "boolean" || typeof obj === "number") return JSON.stringify(obj);
  if (typeof obj === "string") return JSON.stringify(obj);
  if (Array.isArray(obj)) {
    return "[" + obj.map((item) => stableStringify(item)).join(",") + "]";
  }
  if (typeof obj === "object") {
    const keys = Object.keys(obj as Record<string, unknown>).sort();
    const pairs = keys.map(
      (key) =>
        JSON.stringify(key) + ":" + stableStringify((obj as Record<string, unknown>)[key]),
    );
    return "{" + pairs.join(",") + "}";
  }
  return String(obj);
}

/**
 * Compute a deterministic render hash from the project timeline, asset keys, and profile.
 *
 * The hash includes:
 * - All clip timings, ordering, transitions, and effects
 * - All asset references (by R2 object key, not by local path or URL)
 * - Project settings (resolution, fps, sample rate)
 * - Render profile name
 *
 * The hash excludes:
 * - Timestamps (createdAt, modifiedAt)
 * - UI state (selectedClipIds, hoveredClipId, zoom, scroll)
 * - Project name
 *
 * Returns a hex-encoded SHA-256 digest.
 */
export function computeRenderHash(
  project: VideoEditorProject,
  inputAssetKeys: Record<string, string>,
  profile: RenderProfile,
): string {
  const canonical = {
    settings: {
      width: project.settings.width,
      height: project.settings.height,
      fps: project.settings.fps,
      sampleRate: project.settings.sampleRate,
    },
    tracks: canonicalizeTracks(project.timeline.tracks),
    assetKeys: Object.fromEntries(
      Object.entries(inputAssetKeys).sort(([a], [b]) => a.localeCompare(b)),
    ),
    profile,
  };

  return createHash("sha256").update(stableStringify(canonical)).digest("hex");
}

File: python-backend/app/video/render_hash.py
"""Deterministic render hash computation.

Produces a SHA-256 digest from the project timeline, asset keys, and profile.
The hash is used for idempotent rendering: if a render with the same hash
already exists in R2, skip re-rendering.
"""
import hashlib
import json


# Fields to exclude from hash computation (non-deterministic / UI-only)
_EXCLUDED_PROJECT_FIELDS = {"name", "createdAt", "modifiedAt"}
_EXCLUDED_TIMELINE_FIELDS = {"selectedClipIds", "hoveredClipId", "zoom", "scrollLeft", "playbackState", "loopRegion"}


def _canonicalize_clips(clips: list[dict]) -> list[dict]:
    """Sort clips by startTime and extract deterministic fields."""
    sorted_clips = sorted(clips, key=lambda c: c.get("startTime", c.get("startMs", 0)))
    result = []
    for clip in sorted_clips:
        canonical = {
            "assetId": clip.get("assetId"),
            "startTime": clip.get("startTime", clip.get("startMs", 0)),
            "duration": clip.get("duration", clip.get("durationMs", 0)),
            "trimIn": clip.get("trimIn", clip.get("inMs", 0)),
            "trimOut": clip.get("trimOut", clip.get("outMs", 0)),
            "volume": clip.get("volume", 1.0),
            "speed": clip.get("speed", 1.0),
            "effects": clip.get("effects", []),
        }
        if clip.get("inTransition"):
            canonical["inTransition"] = clip["inTransition"]
        if clip.get("transform"):
            canonical["transform"] = clip["transform"]
        if clip.get("textConfig"):
            canonical["textConfig"] = clip["textConfig"]
        if clip.get("transitions"):
            canonical["transitions"] = clip["transitions"]
        result.append(canonical)
    return result


def _canonicalize_tracks(tracks: list[dict]) -> list[dict]:
    """Extract deterministic track data."""
    result = []
    for track in tracks:
        result.append({
            "type": track.get("type"),
            "name": track.get("name"),
            "clips": _canonicalize_clips(track.get("clips", [])),
            "muted": track.get("muted", False),
        })
    return result


def compute_render_hash(
    project: dict,
    input_asset_keys: dict[str, str],
    profile: str,
) -> str:
    """Compute a deterministic render hash from the project timeline, asset keys, and profile.

    The hash includes:
    - All clip timings, ordering, transitions, and effects
    - All asset references (by R2 object key, not by local path or URL)
    - Project settings (resolution, fps, sample rate)
    - Render profile name

    The hash excludes:
    - Timestamps (createdAt, modifiedAt)
    - UI state (selectedClipIds, hoveredClipId, zoom, scroll)
    - Project name

    Returns a hex-encoded SHA-256 digest.
    """
    settings = project.get("settings", {})
    timeline = project.get("timeline", {})

    canonical = {
        "settings": {
            "width": settings.get("width", 1920),
            "height": settings.get("height", 1080),
            "fps": settings.get("fps", 30),
            "sampleRate": settings.get("sampleRate", 48000),
        },
        "tracks": _canonicalize_tracks(timeline.get("tracks", [])),
        "assetKeys": dict(sorted(input_asset_keys.items())),
        "profile": profile,
    }

    json_str = json.dumps(canonical, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(json_str.encode("utf-8")).hexdigest()
